### PR TITLE
DUP: Add a script that "diffs" nix environment generations

### PR DESCRIPTION
**Pull request by shanemikel**
**DUPLICATE** of #22156 (I screwed up the PR with a force push on my fork)

###### Motivation for this change

I found a script like this in matthiasbeyer/nixos-scripts, but it didn't seem to work.  I think it's a useful thing to have, and would like to see it eventually usable via the Nix CLI.  I saw this feature in Guix, and was impressed.  I think it should work for most, but it has a dependency on a coreutils command which isn't present on OSX out of the box (but is present in the nixpkgs coreutils package).

Not sure how people will feel about the style.  Like any code, I try to abstract my Bash, which tends to be hackish and clunky.  Anyway, I think I found a nice way to do something like an import/include...  If y'all don't like it, I can just copy/paste the lib_bash/temp code directly into the script.

### CODE REVIEW

**Comment by shanemikel** [comment]
@LnL7 I've recreated the PR.  If you want to merge it for the immediate benefit of @puffnfresh, I will make a point to address the comments of the original request sometime soon.

**Comment by grahamc** [comment]
👎 on merging it if there are requested changes, just for the benefit of @puffnfresh.

**Comment by shanemikel** [comment]
I think I've satisfied the most important requests from #22156 now, @grahamc

**Comment by shanemikel** [comment]
@grahamc do you maintain the -1?

**Comment by grahamc** [comment]
My -1 isn't specific to anything, if @LnL7 is ready to merge that's fine.
If not, I can review on Monday.

File: maintainers/scripts/nix-diff.sh
Comment (LnL7) [inline maintainers/scripts/nix-diff.sh]: This relies on env-manifest.nix, I think `nix-env-diff` or something would a better name.

File: maintainers/scripts/nix-diff.sh
Comment (LnL7) [inline maintainers/scripts/nix-diff.sh]: I'm not sure about putting this in a separate file.

File: maintainers/scripts/nix-diff.sh
Comment (LnL7) [inline maintainers/scripts/nix-diff.sh]: Perhaps this could default to `$(expr $OPT_TO - 1)`?

**Comment by shanemikel** [comment]
@LnL7 I've made a pretty big revision to the script's behavior.
It now has options to do log-style diffs and an option to more closely mirror the behavior of the guix command that inspired it.

**Comment by shanemikel** [comment]
@LnL7 I don't know what happened to your comment, but I found it in my email...

I've changed the default parent to the newest one older than current.  I didn't use (child-1), rather I compute the most recent one based on what's actually still around.

**Comment by shanemikel** [comment]
Okay, I'm gonna stop changing it for now, unless requested

**Comment by shanemikel** [comment]
What's the best way to verify that ALL the required commands are listed in the shebang for ALL supported platforms?

**Comment by shanemikel** [comment]
@dtzWill would you mind elaborating on what diffing NixOS generations should look like?

**Comment by shanemikel** [comment]
@LnL7 Ready for this one?

**Comment by dtzWill** [comment]
@shanemikel sorry for delay.  No particular vision, hopefully pointing this at ``nix/var/nix/profiles/system` does most of what I had in mind anyway modulo some integration with nixos-rebuild but that is less important.

([original suggestion on earlier version of this PR](https://github.com/NixOS/nixpkgs/pull/22156#issuecomment-275677349))

**Comment by shanemikel** [comment]
@dtzWill 
I've recently added the option to specify a profile manually.

Run it and check that it works for your use case before it's merged.

**Comment by dtzWill** [comment]
LGTM, thanks! :+1:

**Comment by shanemikel** [comment]
Okay, @LnL7, look good?

**Review by LnL7 — APPROVED** [review]

**Comment by LnL7** [comment]
Can you squash this.

**Comment by shanemikel** [comment]
Do I need to create a new PR, or can I force push?

**Comment by LnL7** [comment]
Force pushing is fine.

**Comment by shanemikel** [comment]
Done

**Comment by edolstra** [comment]
I'm concerned `maintainers/scripts` is turning into a dumping ground for random scripts. This script does not even have anything to do with Nix maintenance (it's a general end-user command)!

**Comment by LnL7** [comment]
I didn't really think about that.

@shanemikel In that case  would suggest you create a repo for this and package it.

**Comment by shanemikel** [comment]
There's been much discussion on adding features and generally improving the CLI.

Should we create another directory for these sorts of scripts?  Eventually they could be rewritten into a proper integrated tool?

**Comment by edolstra** [comment]
For discoverability it's best to add it as a package underneath `pkgs/tools/package-management` (like e.g. `nix-repl`). Generally package sources should not be kept in the Nixpkgs repo, but this is fairly small and it's already in the repo anyway...

**Pull request merged**
